Mathlib/Geometry/Convex/Cone/Basic.lean

```diff
@@ -382,6 +382,101 @@ instance instAddCommSemigroup : AddCommSemigroup (ConvexCone R M) where
   add_assoc _ _ _ := SetLike.coe_injective <| add_assoc _ _ _
   add_comm _ _ := SetLike.coe_injective <| add_comm _ _
 
+section Generating
+
+/-- A convex cone `C` is generating if its linear span is the entire `R`-module `M`.
+
+`IsGenerating` is equivalent to `IsReproducing` modulo some conditions.
+See `IsReproducing.isGenerating` and `IsGenerating.isReproducing` for details. -/
+@[simp] def IsGenerating (C : ConvexCone R M) : Prop :=
+  Submodule.span R (C : Set M) = ⊤
+
+/-- The whole `R`-module `M` (viewed as the top convex cone) is generating. -/
+theorem isGenerating_top : (⊤ : ConvexCone R M).IsGenerating := by
+  simp
+
+/-- A convex cone containing a generating cone is also a generating cone. -/
+theorem IsGenerating.mono (h : C₁ ≤ C₂) (hgen : C₁.IsGenerating) : C₂.IsGenerating := by
+  rw [IsGenerating, ← top_le_iff] at hgen ⊢
+  exact hgen.trans (Submodule.span_mono h)
+
+/-- A convex cone is reproducing if its set of element differences equals the entire module,
+i.e., every element of `M` can be written as a difference of two elements of `C`.
+
+This is an alternative notion of generating (`IsGenerating`) which is equivalent in many cases.
+Note: requires additive inverses and hence `[AddCommGroup M]` in place of `[AddCommMonoid M]`. -/
+def IsReproducing {R : Type*} {M : Type*} [Semiring R] [PartialOrder R] [AddCommGroup M]
+    [Module R M] (C : ConvexCone R M) : Prop :=
+  (C : Set M) - (C : Set M) = Set.univ
+
+/-- A reproducing cone is generating.
+
+Note: this direction requires `Ring R` for subtraction in the span submodule, but does not require
+an ordered ring structure or nonemptiness of `C`. -/
+theorem IsReproducing.isGenerating {R : Type*} {M : Type*} [Ring R] [PartialOrder R]
+    [AddCommGroup M] [Module R M] {C : ConvexCone R M} (h : C.IsReproducing) :
+    C.IsGenerating := by
+  rw [IsGenerating, eq_top_iff]
+  intro x _
+  rw [IsReproducing, Set.eq_univ_iff_forall] at h
+  obtain ⟨y, hy, z, hz, heq⟩ := Set.mem_sub.mp (h x)
+  rw [← heq]
+  exact sub_mem (Submodule.subset_span hy) (Submodule.subset_span hz)
+
+/-- A generating cone is reproducing.
+
+Note: this direction requires `Ring R` and stronger ordering assumptions on `R` to handle negative
+scalar multiplication, and nonemptiness of the cone to establish zero membership. -/
+theorem IsGenerating.isReproducing {R : Type*} {M : Type*} [Ring R] [LinearOrder R]
+    [AddLeftStrictMono R] [AddCommGroup M] [Module R M] {C : ConvexCone R M}
+    (hne : (C : Set M).Nonempty) (h : C.IsGenerating) :
+    C.IsReproducing := by
```

Comment:
I think it will be more useful to prove this only the the presence of `[Nontrivial M]` and thus avoid demanding the nonempty assumption which is almost always a trivial consequence of being generating.
```suggestion
/-- A generating cone is reproducing. -/
theorem IsGenerating.isReproducing {R M : Type*} [Ring R] [LinearOrder R]
    [AddLeftStrictMono R] [AddCommGroup M] [Module R M] [Nontrivial M]
    {C : ConvexCone R M} (h : C.IsGenerating) :
    C.IsReproducing := by
  have hne : (C : Set M).Nonempty := by rw [Set.nonempty_iff_ne_empty]; aesop
```